daemon/cmd/daemon.go

```diff
@@ -976,6 +934,53 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
 		close(params.CacheStatus)
 	}
 
+	if params.Clientset.IsEnabled() {
+		bootstrapStats.k8sInit.Start()
+		// Errors are handled inside WaitForCRDsToRegister. It will fatal on a
+		// context deadline or if the context has been cancelled, the context's
+		// error will be returned. Otherwise, it succeeded.
+		if !option.Config.DryMode {
+			if err := d.k8sWatcher.WaitForCRDsToRegister(d.ctx); err != nil {
+				return nil, restoredEndpoints, err
+			}
+		}
+
+		// Launch the K8s node watcher so we can start receiving node events.
+		// Launching the k8s node watcher at this stage will prevent all agents
```

Comment:
Seems dangerous to push this so much later. It might lead to much more traffic to kube-apiserver. How about just moving `WGAgent.Init` + `NodeManager.Subscribe` to before the `k8sInit` step?

Reply:
There is one further complication with this: `WGAgent.Init` also requires `WaitForNodeInformation`, because we need the local node information to determine if the local node opts out of node to node encryption.

Reply:
I have to correct myself. It seems that encryption-opt-out is determined here:
https://github.com/cilium/cilium/blob/433a350ae022889750c27365604f52bae07290c4/daemon/cmd/local_node_init.go#L131 - which is performed via the local K8s Node resource, not `WaitForNodeInformation`.

In that case, it might be fine to move to `WGAgent.Init` earlier in the initialization

Reply:
> How about just moving WGAgent.Init + NodeManager.Subscribe to before the k8sInit step?

Moving the `if params.WGAgent != nil && option.Config.EnableWireguard {` block directly after `bootstrapStats.fqdn.End(true)` does not fix the issue. I think the reason is that the WGAgent is initialized and subscribes to the nodemanager, but it only receives the other nodes when the K8sSubsystem is started. Or more explicitly the watcher for the CiliumNode resource (see: https://github.com/edgelesssys/cilium/blob/fix/wireguard/control-plane-n2n/pkg/k8s/watchers/watcher.go#L455 and https://github.com/edgelesssys/cilium/blob/fix/wireguard/control-plane-n2n/pkg/k8s/watchers/watcher.go#L533).

Reply:
Subscribing to the CiliumNode resource unfortunately won't work beacause that would break clustermesh. WireGuard supports clustermesh, and clustermesh nodes are only available via NodeMangager - CiliumNodes only give you the keys for the local cluster

Reply:
The trace for new nodes currently is
```golang
func (k *K8sWatcher) enableK8sWatchers(ctx context.Context, resourceNames []string) error {
  go k.ciliumNodeInit(k.clientset, asyncControllers)
    k.nodeDiscoverManager.NodeUpdated(n)
      n.Manager.NodeUpdated(node)
        func (m *manager) NodeUpdated(n nodeTypes.Node) {
          func (a *Agent) NodeUpdate(_, newNode nodeTypes.Node) error {
```
I don't want to change that. 

The ciliumNode resource (which already triggers the addition to the WGAgent) is a "start" CRD (see https://github.com/edgelesssys/cilium/blob/fix/wireguard/control-plane-n2n/pkg/k8s/watchers/watcher.go#L455). This means that there is explicitly no need to wait, otherwise it should be after `k.nodeDiscoverManager.WaitForLocalNodeInit()` (see: https://github.com/edgelesssys/cilium/blob/fix/wireguard/control-plane-n2n/pkg/k8s/watchers/watcher.go#L545) and therefore be tagged with `afterNodeInit`.